python/dgl/dataloading/graphsaint.py

```diff
@@ -0,0 +1,128 @@
+    output_device : device, optional
+        The device of the output subgraphs.
+
+    Examples
+    --------
+
+    >>> import torch
+    >>> from dgl.dataloading import SAINTSampler, DataLoader
+    >>> num_iters = 1000
+    >>> sampler = SAINTSampler(mode='node', budget=6000)
+    >>> dataloader = DataLoader(g, torch.arange(num_iters), sampler, num_workers=4)
+    >>> for subg in dataloader:
+    ...     train_on(subg)
+    """
+    def __init__(self, mode, budget, prefetch_ndata=None,
+                 prefetch_edata=None, output_device='cpu'):
+        super().__init__()
+        self.budget = budget
+        if mode == 'node':
+            self.sampler = self.node_sampler
+        elif mode == 'edge':
+            self.sampler = self.edge_sampler
+        elif mode == 'walk':
+            self.sampler = self.walk_sampler
+        else:
+            raise DGLError(f"Expect mode to be 'node', 'edge' or 'walk', got {mode}.")
+
+        self.prefetch_ndata = prefetch_ndata or []
+        self.prefetch_edata = prefetch_edata or []
+        self.output_device = output_device
+
+    def node_sampler(self, g):
+        """Node ID sampler for random node sampler"""
+        # Alternatively, this can be realized by uniformly sampling an edge subset,
+        # and then take the src node of the sampled edges. However, the number of edges
+        # is typically much larger than the number of nodes.
+        prob = g.out_degrees().float().clamp(min=1)
+        return torch.multinomial(prob, num_samples=self.budget, replacement=True)
+
+    def edge_sampler(self, g):
+        """Node ID sampler for random edge sampler"""
+        src, dst = g.edges()
+        in_deg = g.in_degrees().float().clamp(min=1)
+        out_deg = g.out_degrees().float().clamp(min=1)
+        # We can reduce the sample space by half if graphs are always symmetric.
+        prob = 1. / in_deg[dst.long()] + 1. / out_deg[src.long()]
+        prob /= prob.sum()
+        sampled_edges = torch.unique(choice(len(prob), size=self.budget, prob=prob))
+        sampled_nodes = torch.cat([src[sampled_edges], dst[sampled_edges]])
+        return sampled_nodes
+
+    def walk_sampler(self, g):
+        """Node ID sampler for random walk sampler"""
+        num_roots, walk_length = self.budget
+        sampled_roots = torch.randint(0, g.num_nodes(), (num_roots,))
+        traces, types = random_walk(g, nodes=sampled_roots, length=walk_length)
+        sampled_nodes, _, _, _ = pack_traces(traces, types)
+        return sampled_nodes
+
+    def sample(self, g, indices):
+        """Sampling function
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph to sample from.
+        indices : Tensor
+            Placeholder not used.
+
+        Returns
+        -------
+        DGLGraph
+            The sampled subgraph.
+        """
+        node_ids = self.sampler(g).unique().type(g.idtype)
```

Comment:
I will put `unique` and `type` in the `xx_sampler` methods.

Reply:
Done